oracle-token/src/lib.rs

```diff
@@ -0,0 +1,1170 @@
+            Error::CalculationFailure => info!("Error: CalculationFailure"),
+        }
+    }
+}
+
+pub type Ratio = (u64, u64);
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OracleValues {
+    latest: u64,
+    values: [(Slot, Ratio); 32],
+}
+impl OracleValues {
+    fn from_account_info(_info: &AccountInfo) -> Result<Self, ProgramError> {
+        unimplemented!();
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OracleToken {
+    self_pubkey: Pubkey,
+    authority: Pubkey,
+    token_program: Pubkey,
+    oracle_values: Pubkey,
+    colleteral: Pubkey,
+    pool_mint: Pubkey,
+    synthetic_mint: Pubkey,
+    pool_supply: u64,
+    synthetic_supply: u64,
+    info: OracleTokenInfo,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Deposit {
+    slot: Slot,
+    amount: u64,
+    user: Pubkey,
+}
+
+impl Deposit {
+    fn calculate(
+        &self,
+        clock: &Clock,
+        info: &OracleTokenInfo,
+        pool: u64,
+        colleteral: u64,
+        synthetic: u64,
+        oracle_values: &OracleValues,
+    ) -> Option<u64> {
+        let (slot, price) = oracle_values.values.get(oracle_values.latest as usize)?;
+        if *slot < self.slot + info.min_slots && clock.slot < self.slot + info.max_slots {
+            return None;
+        }
+        let obligation = synthetic as f64 * price.0 as f64 / price.1 as f64;
+        let base = colleteral as f64 - obligation;
+        let percentage = self.amount as f64 / base;
+        let amount = percentage * pool as f64;
+        Some(amount as u64)
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Withdraw {
+    slot: Slot,
+    amount: u64,
+    user: Pubkey,
+}
+
+impl Withdraw {
+    fn calculate(
+        &self,
+        clock: &Clock,
+        info: &OracleTokenInfo,
+        pool: u64,
+        colleteral: u64,
+        synthetic: u64,
+        oracle_values: &OracleValues,
+    ) -> Option<u64> {
+        let (slot, price) = oracle_values.values.get(oracle_values.latest as usize)?;
+        if *slot < self.slot + info.min_slots && clock.slot < self.slot + info.max_slots {
+            return None;
+        }
+        let obligation = synthetic as f64 * price.0 as f64 / price.1 as f64;
+        let percentage = self.amount as f64 / pool as f64;
+        let amount = (colleteral as f64 - obligation) * percentage;
+        Some(amount as u64)
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Swap {
+    slot: Slot,
+    amount: u64,
+    user: Pubkey,
+    base_to_synthetic: bool,
+}
+
+impl Swap {
+    fn calculate(
+        &self,
+        clock: &Clock,
+        info: &OracleTokenInfo,
+        pool: u64,
+        colleteral: u64,
+        synthetic: u64,
+        oracle_values: &OracleValues,
+    ) -> Option<u64> {
+        let (slot, price) = oracle_values.values.get(oracle_values.latest as usize)?;
```

Comment:
Assuming a pool with 1:1 price
```
Pool supply: 100
Collateral: 200
Synth: 100
```

A user wants to swap 100 synth to collateral/base.

As in the comment before, `synthetic` would already be `100-100=0` making obligation `0` and percentage `100/100=1`, making amount `collateral=200`.

Applying the same solution as in the previous comment this would yield a conversion of 100->100 (before fees) which is expected.

Reply:
I should stop coding at night ;).  Obviously unit tests are still TBD